features/netsocket/UDPSocket.cpp

```diff
@@ -85,14 +89,32 @@ nsapi_size_or_error_t UDPSocket::sendto(const SocketAddress &address, const void
         }
     }
 
+    _writers--;
+    if (!_socket || !_writers) {
+         _event_flag.set(FINISHED_FLAG);
+    }
     _lock.unlock();
     return ret;
 }
 
+nsapi_size_or_error_t UDPSocket::send(const void *data, nsapi_size_t size)
+{
+    if (!_remote_peer)
+        return NSAPI_ERROR_NO_ADDRESS;
+    return sendto(_remote_peer, data, size);
+}
+
 nsapi_size_or_error_t UDPSocket::recvfrom(SocketAddress *address, void *buffer, nsapi_size_t size)
 {
     _lock.lock();
     nsapi_size_or_error_t ret;
+    SocketAddress ignored;
+
+    if (!address) {
```

Comment:
Possible refinement - `connect()` is supposed to act as a filter on received packets, as well as be the destination for `send()`.

Maybe we should have a `if (recv >= 0 && _remote_peer && address != _remote_peer) continue` in the loop?

Reply:
I'm OK with that idea, but are we allowed to trash the buffers of caller, if we don't provide any data for them?
For example:
Socket is set to non-blocking mode and connected
Packet comes in
sigio() happens
Application calls recvfrom()
packet is not from the _peer_address, so continue happens
timeout
return WOULD_BLOCK

Now we are telling that we did not receive anything, but we did write the incoming packet to callers buffer, and address is recorded in the provided storage.

Reply:
Also, POSIX says `For SOCK_DGRAM sockets, the peer address identifies where all datagrams are sent on subsequent send() functions, and limits the remote sender for subsequent recv() functions. `

So should we filter only on `recv()` or both `recv()` and `recvfrom()`?

Reply:
>  we did write the incoming packet to callers buffer

Good point. Can the caller assume buffer is untouched if error is returned?  It certainly would be in a real POSIX system, cos of the way system calls works, but I'm not sure it's a useful guarantee for applications, whereas the ability to use input buffer as "scratch" seems like it's potentially generally useful for drivers/stacks.

I'd be inclined to maybe add a little API note that buffer and address may be modified even if the call ends up returning an error. 

> Also, POSIX says

I saw that, but concluded it was poor wording - they must mean `recv*`.  The equivalence between `recv` and `recvfrom` forces it to affect both, I believe:

> `The recv() function is equivalent to recvfrom() with a zero address_len argument`

I guess the `connect` specifically makes `recv()` useful, in that you can assume the remote address, but the filtering must apply generally. (In a real stack the filtering would happen on entry to the socket buffer, not exit, so call chosen to read the buffer can't be a factor).